src/stan/model/indexing/rvalue_cl.hpp

```diff
@@ -0,0 +1,333 @@
+          require_not_same_t<RowIndex, math::matrix_cl<int>>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex row_index) {
+  int rows = expr.rows();
+  internal::index_check(row_index, name, rows);
+  auto res_vari
+      = expr.vi_->index(internal::cl_row_index(row_index, expr.rows(), name),
+                        math::col_index(-1, expr.cols()));
+  return math::var_value<value_type_t<decltype(res_vari)>>(
+      new decltype(res_vari)(std::move(res_vari)));
+}
+
+/**
+ * Index a rev kernel generator expression with two (non-multi) indices.
+ *
+ * @tparam Expr type of the expression
+ * @tparam RowIndex type of row index
+ * @tparam ColIndex type of column index
+ * @param expr a prim kernel generator expression to index
+ * @param name name of value being indexed (if named, otherwise an empty string)
+ * @param row_index row index
+ * @param col_index column index
+ * @return result of indexing
+ */
+template <typename Expr, typename RowIndex, typename ColIndex,
+          require_rev_kernel_expression_t<Expr>* = nullptr,
+          require_not_same_t<RowIndex, math::matrix_cl<int>>* = nullptr,
+          require_not_same_t<ColIndex, math::matrix_cl<int>>* = nullptr,
+          require_any_not_same_t<RowIndex, ColIndex, index_uni>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex row_index,
+                   const ColIndex col_index) {
+  int rows = expr.rows();
+  int cols = expr.cols();
+  internal::index_check(row_index, name, rows);
+  internal::index_check(col_index, name, cols);
+
+  auto res_vari
+      = expr.vi_->index(internal::cl_row_index(row_index, rows, name),
+                        internal::cl_col_index(col_index, expr.cols(), name));
+  return math::var_value<value_type_t<decltype(res_vari)>>(
+      new decltype(res_vari)(std::move(res_vari)));
+}
+
+/**
+ * Index a rev kernel generator expression with two uni indices.
+ *
+ * @tparam Expr type of the expression
+ * @param expr a prim kernel generator expression to index
+ * @param name name of value being indexed (if named, otherwise an empty string)
+ * @param row_index row index
+ * @param col_index column index
+ * @return result of indexing
+ */
+template <typename Expr, require_rev_kernel_expression_t<Expr>* = nullptr>
+inline math::var rvalue(Expr&& expr, const char* name,
+                        const index_uni row_index, const index_uni col_index) {
+  using Val = stan::value_type_t<stan::value_type_t<Expr>>;
+  math::check_range("uni indexing", name, expr.rows(), row_index.n_);
+  math::check_range("uni indexing", name, expr.cols(), col_index.n_);
+  cl::CommandQueue queue = stan::math::opencl_context.queue();
+  Val res;
+  try {
+    cl::Event copy_event;
+    queue.enqueueReadBuffer(
+        expr.val().buffer(), true,
+        sizeof(Val) * (row_index.n_ - 1 + (col_index.n_ - 1) * expr.rows()),
+        sizeof(Val), &res, &expr.val().write_events(), &copy_event);
+    copy_event.wait();
```

Comment:
Do these expressions need evaluated before we write to their buffer?

Reply:
Expressions in general do not have buffers. Only matrices do. So we need to evaluate the expression to get a matrix. Also this is reading the buffer, not writing it.